libgeopmd/src/PowercapSysfsDriver.cpp

```diff
@@ -0,0 +1,196 @@
+        if (policy_file.find("intel-rapl:") != 0) {
+            continue;
+        }
+
+        std::ostringstream oss;
+        oss << powercap_directory << "/" << policy_file << "/name";
+        auto name_path = oss.str();
+
+        int name_fd = open(name_path.c_str(), O_RDONLY);
+        if (name_fd == -1) {
+            throw geopm::Exception("PowercapSysfsDriver failed to open " + name_path,
+                                   errno, __FILE__, __LINE__);
+        }
+        char name_buf[geopm::SysfsDriver::M_IO_BUFFER_SIZE] = {};
+        int read_bytes = pread(name_fd, name_buf, sizeof(name_buf) - 1, 0);
+        close(name_fd);
+        if (read_bytes < 0) {
+            throw geopm::Exception("PowercapSysfsDriver failed to read " + name_path,
+                                   errno, __FILE__, __LINE__);
+        }
+        if (read_bytes == sizeof(name_buf) - 1) {
+            throw geopm::Exception("PowercapSysfsDriver truncated read from " + name_path,
+                                   errno, __FILE__, __LINE__);
+        }
+
+        std::string name(name_buf);
+        if (!name.empty() && name.back() == '\n') {
+            name.pop_back();
+        }
+        result.emplace(name, policy_file);
+    }
+
+    return result;
+}
+
+namespace geopm
+{
+    const std::string powercap_sysfs_json(void);
+
+    PowercapSysfsDriver::PowercapSysfsDriver()
+        : PowercapSysfsDriver(POWERCAP_DIRECTORY)
+    {
+    }
+
+    PowercapSysfsDriver::PowercapSysfsDriver(const std::string &powercap_directory)
+        : M_PROPERTIES{SysfsDriver::parse_properties_json(plugin_name(), powercap_sysfs_json())}
+        , M_POWERCAP_RESOURCE_BY_NAME(load_powercap_resource_by_name(powercap_directory))
+        , m_powercap_directory(powercap_directory)
+    {
+    }
+
+    int PowercapSysfsDriver::domain_type(const std::string &name) const
+    {
+        if (M_PROPERTIES.find(name) == M_PROPERTIES.end()) {
+            throw Exception("PowercapSysfsDriver: unknown name: " + name,
+                            GEOPM_ERROR_RUNTIME, __FILE__, __LINE__);
+        }
+        return GEOPM_DOMAIN_PACKAGE;
+    }
+
+    static std::string name_to_resource_key(const std::string &name, int domain_idx)
+    {
+        std::string resource_key;
+        if (name.find("POWERCAP::CPU_") == 0) {
+            resource_key = "package-" + std::to_string(domain_idx);
+        }
+        else if (name.find("POWERCAP::DRAM_") == 0) {
+            resource_key = "dram";
+        }
+        else {
+            throw Exception("PowercapSysfsDriver: unknown name: " + name,
+                            GEOPM_ERROR_RUNTIME, __FILE__, __LINE__);
+        }
+        return resource_key;
+    }
+
+    static std::string name_to_property_key(const std::string &name)
+    {
+        std::string property_key;
+        auto substr_pos = name.find("POWERCAP::CPU_");
+        if (substr_pos == std::string::npos) {
+            substr_pos = name.find("POWERCAP::DRAM_");
```

Comment:
Minor suggestion: seeing these literal strings (passed to the `find` function) are used in more than one place, I would recommend putting them in a constant.